common/DynamicLibrary.cpp

```diff
@@ -0,0 +1,166 @@
+/*  PCSX2 - PS2 Emulator for PCs
+ *  Copyright (C) 2021 PCSX2 Dev Team
+ *
+ *  PCSX2 is free software: you can redistribute it and/or modify it under the terms
+ *  of the GNU Lesser General Public License as published by the Free Software Found-
+ *  ation, either version 3 of the License, or (at your option) any later version.
+ *
+ *  PCSX2 is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with PCSX2.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef _WIN32
+#include <Windows.h>
+#include <wil/com.h>
+#else
+#include <dlfcn.h>
+#endif
+
+#include <fmt/core.h>
+#include "common/DynamicLibrary.h"
+
+namespace Common
+{
+#if defined(_WIN32)
+	struct DynamicLibrary::Impl
+	{
+		static constexpr const char* ExtName()
+		{
+			return "dll";
+		}
+
+		bool Open(std::string_view name)
+		{
+			constexpr DWORD flags =
+				LOAD_LIBRARY_SEARCH_APPLICATION_DIR |
+				LOAD_LIBRARY_SEARCH_SYSTEM32;
+
+			m_handle.reset(LoadLibraryExA(name.data(), nullptr, flags));
+
+			return IsOpen();
+		}
+
+		void Close()
+		{
+			m_handle.reset();
+		}
+
+		void* GetSymbolByName(std::string_view name)
+		{
+			auto sym = GetProcAddress(m_handle.get(), name.data());
+
+			return reinterpret_cast<void*>(sym);
+		}
+
+		void* GetSymbolByOrdinal(const int ordinal)
+		{
+			auto sym = GetProcAddress(m_handle.get(), reinterpret_cast<LPCSTR>(ordinal));
+
+			return reinterpret_cast<void*>(sym);
+		}
+
+		bool IsOpen() const
+		{
+			return m_handle != nullptr;
+		}
+
+		wil::unique_hmodule m_handle{ nullptr };
+	};
+#else
+	struct DynamicLibrary::Impl
+	{
+		static constexpr const char* ExtName()
+		{
+#if defined(__APPLE__)
+			return "dylib";
+#else
+			return "so";
+#endif
+		}
+
+		bool Open(std::string_view name)
+		{
+			m_handle = dlopen(name.data(), RTLD_NOW);
+
+			return IsOpen();
+		}
+
+		void Close()
+		{
+			if (m_handle)
+				dlclose(m_handle);
+		}
+
+		void* GetSymbolByName(std::string_view name)
+		{
+			return dlsym(m_handle, name.data());
+		}
+
+		void* GetSymbolByOrdinal(const int /*ordinal*/)
+		{
+			return nullptr;
```

Comment:
If it's only used in ifdef'd Windows code, thoughts on leaving it out of the UNIX API completely?  That way attempting to use it would be a compiler error instead of a runtime one?